src/main/java/games/strategy/engine/framework/map/download/MapDownloadProgressListener.java

```diff
@@ -0,0 +1,59 @@
+package games.strategy.engine.framework.map.download;
+
+import java.util.Optional;
+
+import javax.swing.JProgressBar;
+import javax.swing.SwingUtilities;
+
+/**
+ * A listener of map download progress events that updates the associated controls in the UI.
+ *
+ * <p>
+ * Instances of this class are thread safe.
+ * </p>
+ */
+final class MapDownloadProgressListener {
```

Comment:
This class was added when I changed how download length was modeled from `Integer` to `Long`.

Previously, the progress bar's maximum range was set to the download length.  When I switched the download length type to `Long`, that value was no longer compatible without casting, and casting just felt dirty (even though I doubt we'll ever have a download larger than 2 GiB :smiley:).

The solution I chose was to fix the progress bar range to [0,100] and to have the "progress updater" (this class) map the (`Long`) download progress into this (`Integer`) range (see implementation below in `downloadUpdated()`).

This class is also where all the indeterminate progress bar stuff is managed.